### PR TITLE
[Inductor] fix addmm fusion check

### LINKED ISSUE #121253: [torch.compile][freezing] `addmm` results in a significant decrease in accuracy
### 🐛 Describe the bug

When using `TORCHINDUCTOR_FREEZING=1`, `torch.compile` will result in a significant decrease in accuracy for `addmm`

```py
import torch

torch.manual_seed(42)

class Model(torch.nn.Module):

    def __init__(self, weight, bias):
        super(Model, self).__init__()
        self.weight = weight
        self.bias = bias

    def forward(self, x):
        return torch.addmm(self.bias, x, self.weight, alpha=1.0, beta=0.1)

weight = 1
bias = 1
func = Model(torch.randn(64, 64), torch.randn(64)).to('cpu')

x = torch.randn(1, 64)

# x = x.to(dtype=torch.float64)
# func.weight = func.weight.to(dtype=torch.float64)
# func.bias = func.bias.to(dtype=torch.float64)

with torch.no_grad():
    func1 = torch.compile(func)
    print(func1(x.clone()))
    # tensor([[  5.5041,   8.0136,   9.7231,   4.3172,  -4.7227,   7.4632, -10.4915,
    #       -0.5536, -15.7986,   8.6184,   3.1239,  -1.8854,   7.8321,   3.1380,
    #        0.0261,   2.4693,  -0.1138,  -5.1792,  -4.9836,  -7.1234,   3.7069,
    #        6.9476,  22.1048,  -3.7576,  12.4651, -10.9513,   7.8899,   4.4365,
    #        0.0233,  12.4531,   5.4630,   2.7396,  15.0662, -11.8093,  -7.6297,
    #        1.5935,  -1.6402,  11.1157,  -2.1696, -17.6288,   4.0042,   2.1122,
    #      -16.2381,  19.1538,   0.2264,   0.5900,  -7.5213,  14.8976,  -6.1498,
    #      -11.5165,  -6.5250,   8.4473,   3.2056,   5.9807,  -8.0135,  -5.0815,
    #       -3.4325,  -2.4708,   3.6546,   1.3412,  -8.9707,  21.8540,   0.4926,
    #       -3.8654]])

    print(func(x.clone()))
    # tensor([[  5.9718,   6.3465,   8.0702,   2.4505,  -4.0591,   8.1550, -10.4454,
    #       -1.9923, -15.9896,   7.6231,   1.9432,  -2.2672,   8.2752,   1.6473,
    #       -0.3439,   2.6879,  -0.8909,  -3.8750,  -5.2665,  -6.2186,   4.9161,
    #        5.8066,  23.2692,  -3.0903,  12.7620, -11.2484,   7.0066,   5.7786,
    #       -0.4614,  11.2506,   5.9704,   2.1425,  14.6779, -11.1137,  -6.9070,
    #        1.1754,  -1.4864,   8.6796,  -2.7658, -17.0402,   3.3492,   2.0289,
    #      -16.0763,  18.5234,   1.3519,  -0.2281,  -7.3793,  15.2927,  -6.8148,
    #      -11.0951,  -5.0884,   9.9478,   2.9006,   5.6432,  -6.8172,  -5.5889,
    #       -4.1571,  -3.1412,   3.8111,   0.3441,  -9.4339,  21.0498,   1.8529,
    #       -3.0991]])
```

I tried using `float64` and the result is indeed the same as naive execution with `float32`. Thus, I think there would be an issue in the `freezing` optimization, probably the constant folding for `addmm`

This could be related to #121175

### Versions

```
Versions
PyTorch version: 2.3.0.dev20240301+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.0 (default, Nov 15 2020, 14:28:56) [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-6.5.0-21-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 24
On-line CPU(s) list: 0-23
Vendor ID: GenuineIntel
Model name: 12th Gen Intel(R) Core(TM) i9-12900K
CPU family: 6
Model: 151
Thread(s) per core: 2
Core(s) per socket: 16
Socket(s): 1
Stepping: 2
CPU max MHz: 5200.0000
CPU min MHz: 800.0000
BogoMIPS: 6374.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 640 KiB (16 instances)
L1i cache: 768 KiB (16 instances)
L2 cache: 14 MiB (10 instances)
L3 cache: 30 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240301+cu118
[pip3] torchaudio==2.2.0.dev20240301+cu118
[pip3] torchvision==0.18.0.dev20240301+cu118
[conda] numpy 1.26.4 pypi_0 pypi
[conda] pytorch-triton 3.0.0+901819d2b6 pypi_0 pypi
[conda] torch 2.3.0.dev20240301+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20240301+cu118 pypi_0 pypi
[conda] torchvision 0.18.0.dev20240301+cu118 pypi_0 pypi
```

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang

### PR DESCRIPTION
**Pull request by Valentine233**
Fixes #121253.

To avoid functional issue, disable pattern match for `addmm` when `beta!=1 or 0` or `alpha!=1`, as either `mkl_linear` or `mkldnn_linear` doesn't accept `beta` or `alpha` as parameters.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 1]
```diff
@@ -910,6 +910,11 @@ def _is_packable_linear(match):
         Check if the node is supported for MKLDNN linear.
         """
         linear_node = match.output_node()
+        # mkldnn linear only supports beta=1 and alpha=1
+        # beta_idx=3, alpha_idx=4
+        if linear_node.target == aten.addmm.default:
+            if match.args[3] != 1.0 or match.args[4] != 1.0:
```
Comment (jgong5) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:916]: when beta is 0, it is the case without bias. mkldnn linear still supports it, right?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/mkldnn_fusion.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/mkldnn_fusion.py:916]: Maybe using `KeywordArg("beta")` like https://github.com/pytorch/pytorch/blob/5f601a41e0a8c91ecf7ca5e4b95d752166ed9093/torch/_inductor/fx_passes/mkldnn_fusion.py#L972. It should be more readable.

Comment (Valentine233) [reply torch/_inductor/fx_passes/mkldnn_fusion.py]: @jgong5 You are right. I have added the support of beta=0.

Comment (Valentine233) [reply torch/_inductor/fx_passes/mkldnn_fusion.py]: @leslie-fang-intel Thanks, Leslie. I tried to use `KeywordArg`, but it could not be used by `beta` and `alpha`, as they are optional.

Comment (Valentine233) [reply torch/_inductor/fx_passes/mkldnn_fusion.py]: @leslie-fang-intel Thanks! Modified to `KeyWordArg`.

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 2]
```diff
@@ -2127,6 +2127,32 @@ def forward(self, x):
...
+        for beta, alpha in zip([1.0, 0.1, 0.0], [1.0, 0.1, 1.0]):
+            mod = Mod(torch.randn(64, 64), torch.randn(64), beta, alpha).eval()
+            with torch.no_grad():
+                x = torch.randn(1, 64)
+                include_ops = []
+                exclude_ops = []
+                if (beta != 1.0 and beta != 0.0) or alpha != 1.0:
+                    exclude_ops = ["mkldnn._linear_pointwise.default"]
+                else:
+                    exclude_ops = ["mkldnn._linear_pointwise.default"]
```
Comment (leslie-fang-intel) [inline test/inductor/test_mkldnn_pattern_matcher.py:2153]: Why same `exclude_ops`? I think for FP32, we are using `torch.ops.mkl._mkl_linear`. Did you mean check of it?

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 3]
```diff
@@ -910,6 +910,18 @@ def _is_packable_linear(match):
         Check if the node is supported for MKLDNN linear.
         """
         linear_node = match.output_node()
+        # mkldnn linear only supports beta=1or0 and alpha=1
+        if linear_node.target == aten.addmm.default:
+            alpha = (
+                None
+                if "alpha" not in linear_node.kwargs
+                else linear_node.kwargs["alpha"]
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:918]: Try `linear_node.kwargs.get("alpha", 1.0)`, it may provide a default value when not exist.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 2 above)
Comment (Valentine233) [reply test/inductor/test_mkldnn_pattern_matcher.py]: Sorry for the bug... Added UT for both FP32 and BF16.

File: torch/_inductor/fx_passes/mkldnn_fusion.py (hunk 3 above)
Comment (Valentine233) [reply torch/_inductor/fx_passes/mkldnn_fusion.py]: Thanks and modified.

**Review by leslie-fang-intel — APPROVED** [review]
LGTM

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]
